lib/improver/cli/ensemble_calibration.py

```diff
@@ -161,6 +161,29 @@ def main(argv=None):
                              'is converted to percentiles, as part of '
                              'converting the input probabilities into '
                              'realizations.')
+    parser.add_argument('--max_iterations', metavar='MAX_ITERATIONS',
+                        type=np.int32, default=1000,
+                        help='The maximum number of iterations allowed '
+                             'until the minimisation has converged to a '
+                             'stable solution. If the maximum number '
+                             'of iterations is reached, but the '
+                             'minimisation has not yet converged to a '
+                             'stable solution, then the available solution '
+                             'is used anyway, and a warning is raised.'
+                             'If the predictor_of_mean is "realizations", '
+                             'then the number of iterations may require '
+                             'increasing, as there will be more coefficients '
+                             'to solve for.')
+    parser.add_argument('--decimals', metavar='DECIMALS',
+                        type=np.int32, default=5,
+                        help='Integer to define how many decimals the inputs '
+                             'to the minimisation will be rounded to. This is '
+                             'for the purpose of obtaining consistent output '
+                             'from the minimisation across different packages '
+                             'and processors, where differences at the '
+                             'machine precision level will cause the '
+                             'minimisation to go in different directions and '
```

Comment:
I would guess that 'go in different directions' depends on the exact nature of the landscape of the function to be minimised. I'd perhaps just leave this as 'minimisation to converge to different solutions'.

Reply:
(similarly for the similar help strings/docstrings etc)

Reply:
I think we probably want to use the `tol` option to `minimize` though, rather than rounding all of the inputs.

Reply:
After taking a fresh look at this, I've found that just converting the inputs to the minimisation to be float64, rather than converting the inputs to float32 and rounding the inputs, is better from the perspective of obtaining more stable solutions from the unit tests, at least for the different machines that I've run the tests on. There are still occasionally some differences in the coefficients within the unit tests when iterating to convergence, but this is limited to the the ~5th decimal place, rather than causing markedly different coefficients to be found. As a result of this, I've removed the `decimals` argument as it is no longer required.

Reply:
That's great news. Sounds to me like the algorithm was numerically unstable with the reduced precision.

My recommendations would be:
   * use float64
   * don't round inputs, but instead adjust the tolerance
   * use array almost equal in unit tests

(For almost all iterative real world applications of iterative optimization (that I've seen), you typically almost always need float64.)

Reply:
Following our conversation earlier, I've now adjusted the unit tests that are using `assertArrayAlmostEqual` to use a lower precision version where arrays are compared to 4 decimal places only. This is justified based on that the [Nelder-Mead minimisation](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-neldermead.html) has a default tolerance of 0.0001, so the aim is for coefficients to match to a precision of 4 decimal places.